data/blog/umdctf-2023/pokeptcha.md

```diff
@@ -0,0 +1,414 @@
+
+```python
+cwJjVO[3] = eval("var Yulo=['\\x70\\x75\\x73','\\x70\\x6f\\x70','\\x68','\\x70\\x6f\\x70'];var i0M=function(){var qIdHpk=fWu[Yulo[(-(-(177/(59*(91/91)))+(3000/50-1043%62))+7)*1]]();var j9M=fWu[Yulo[91-(26+(42+20))]]();j9M[Yulo[0/(87-118%34)]+Yulo[82%5]](qIdHpk)};i0M");
+```
+
+Decoding this gives us a new operation:
+
+```javascript
+// 3
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    b.push(a)
+}
+```
+
+Which is just appending one value to the other. The next thing we do (after some push and evals) is actually this, where we append to the operations. We continue adding more operations and usually use them right after adding them to the operations array, but after some time I finally had them all:
+
+```javascript
+// 4, pop
+function() {
+    stack.pop()
+}
+
+// 5, mem_write
+function set_mem(value,address){
+    var value_func = () => { return value };
+    memory[address] = value_func
+}
+function(){
+    var address = stack.pop();
+    var value = stack.pop();
+    set_mem(value, address)
+}
+
+// 6, mem_read
+function get_mem(address) {
+    return memory[address]()
+}
+function(){
+    var address = stack.pop();
+    stack.push(get_mem(address))
+}
+
+// 7, sub
+function(){
+    var a=stack.pop();
+    var b=stack.pop();
+    stack.push(a-b)
+}
+
+// 8, or
+function(){
+    stack.push(stack.pop()||stack.pop())
+}
+
+// 9, push_global
+function(){
+    stack.push(global)
+}
+
+// 10, push_null
+function(){
+    stack.push(null)
+}
+
+// 11, goto
+function(){
+    pc=stack.pop()
+}
+
+// 12, call
+function(){
+    var func=stack.pop();
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(func.apply(a,b))
+}
+
+// 13, push_arr
+function(){
+    stack.push([])
+}
+
+// 14, jne
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    var dst = stack.pop();
+    pc = a == b ? pc : dst
+}
+
+// 15, xor
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(a^b)
+}
+
+// 16, mod
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(a%b)
+}
+
+// 17, index
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(a[b])
+}
+
+// 18, add
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(a+b)
+}
+
+// 19, shr
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    stack.push(b>>a)
+}
+
+// 20, jne again?
+function(){
+    var a = stack.pop();
+    var b = stack.pop();
+    var dst = stack.pop();
+    pc = a != b ? dst : pc
+}
+```
+
+Not quite sure about the two `jne` instructions, could be that I switched some variables around and one is actually a `je`, but from now I just printed the operations and didn’t interpret them anymore, so it didn’t matter too much (I just needed to know that there is a conditional jump to some address). With a bit of cleanup I got:
```

Comment:
```suggestion
Not quite sure about the two `jne` instructions—it could be that I switched some variables around and one is actually a `je`. But from now, I just printed the operations and didn’t interpret them anymore, so it didn’t matter too much (I just needed to know that there is a conditional jump to some address). A bit of cleanup resulted in this:
```